### ISSUE TITLE
[PROPOSAL] Addition of a tags Field in Event : OpenLineage.json Schema 

### ISSUE THREAD

**Issue by surisimran** [OpenLineage/OpenLineage#2748]
**Background:**

The OpenLineage schema is designed to capture and track the lifecycle of data processes. While it effectively records various aspects of data lineage, the ability to attach user-defined tags to events can greatly enhance the schema's flexibility and utility. Tags can provide additional context and enable more effective categorization, filtering, and querying of lineage data. This is similar to how tags/properties are used in dbt models, where users can pass tags along with descriptions for datasets or columns.

**Purpose:**

Field Name: tags (or properties if that aligns better with existing schema conventions)
Field Type: Array of Strings (List of key-value pairs if using properties)
Description: This field will allow users to attach arbitrary tags to events. These tags can be used for classification, filtering, and enhancing the context of the data lineage.

**Benefits:**

Enhanced Metadata Capture: Users can add specific tags that provide additional information about the event, such as environment (e.g., production, staging), project name, data sensitivity level, and more.
Improved Classification: Tags enable better classification and organization of lineage events, making it easier to manage and understand large and complex data workflows.
Customizable Context: Users can define their own tags based on their specific needs, allowing for a more flexible and adaptable schema.

**Example Usage:**

```
{
  "eventType": "START",
  "eventTime": "2023-06-01T12:00:00Z",
  "run": {
    "runId": "12345",
    "facets": {
      // Existing facets
    }
  },
  "job": {
    "namespace": "my_namespace",
    "name": "my_job"
  },
  "producer": "my_producer",
  "tags": {
    "environment": "production",
    "project": "data_migration",
    "data_sensitivity": "sensitive",
    "team": "analytics"
  //or like "tags": ["production", "etl", "sensitive_data"]
  }
}

```

**Comment by dolfinus** [comment]
Can this be a one more facet?

**Comment by surisimran** [comment]
Yes, definately it can be a facet

**Comment by mobuchowski** [comment]
First, I think we should not have free-standing field with tags in the top-level event, as event should be purely description of defined core entities rather than be a entity on it's own. 

However, we probably should have facets that allows to tag jobs, datasets and runs. There was some discussion in https://github.com/OpenLineage/OpenLineage/issues/1160 about tagging.

**Comment by surisimran** [comment]
Not only jobs, datasets, runs but also if we can have for columns as well for data sensitivity level, and more. 
Seems like that discussion/PR got closed at that time. Can we please re-check on the same.

**Comment by mobuchowski** [comment]
Yeah, for columns there could be several choices - for example, include this data in SchemaDatasetFacet or ColumnLineageDatasetFacet, or in different, new facet.

**Comment by 7vaibhav7** [comment]
Great, can someone please incorporate this as a facet, I'm not sure on implementation? @mobuchowski @dolfinus  

I was also looking for something similar to have tags both generic on dataset (event level) and column level.
Then I guess a new facet for generic dataset (event level) and for column level in SchemaDatasetFacet or ColumnLineageDatasetFacet can be done.

**Comment by mobuchowski** [comment]
@7vaibhav7 the best way to have this implemented is to do it yourself :) 

It would be nice if you could show the design and how you want to populate the facet fields themselves.

**Comment by 7vaibhav7** [comment]
Ideally, I'm looking for something similar to what @surisimran mentioned above - a kind of key-value pairs may be as a tag field at various levels: event, dataset (input, output), and column. For example:

**Generic Tags in Event:**
```
{
  "eventType": "START",
  "eventTime": "2023-06-01T12:00:00Z",
  "run": {
    "runId": "12345",
    "facets": {
      // Existing facets
    }
  },
  "job": {
    "namespace": "my_namespace",
    "name": "my_job"
  },
  "producer": "my_producer",
  "tags": {
    "environment": "production",
    "project": "data_migration",
    "data_sensitivity": "sensitive",
    "team": "analytics",
    "other": ["dummydata", "dummydata2"]
  }
}
```

Similarly tags keys can be changed in input/output and coulmn facets:

**Input/Output facets:**
```
 "tags": {
    "environment": "production",
    "project": "data_migration",
    "data_sensitivity": "sensitive",
    "team": "analytics",
    "other": ["dummydata", "dummydata2"]
```

**ColumnLineageDatasetFacet addition EXAMPLE:** 
 ```
  {
      "name": "column1",
      "tags": {
        "security_classification": "high",
        "sensitivity": "personal",
        "profilers": ["profiler1", "profiler2"],
        "other": ["example1", "example2"]
      }
    },
    {
      "name": "column2",
      "tags": {
        "security_classification": "medium",
        "sensitivity": "confidential",
        "profilers": ["profiler3", "profiler4"],
        "other": ["example3", "example4"]
      }
    }
```

These examples might not align perfectly with the proper schema, but I hope you understand the examples I'm trying to give. @mobuchowski  @dolfinus 

In this way, tags can be included everywhere at all levels: generic, dataset, and column. The keys can be managed within code statically, or the user can provide them dynamically as they wish.

**Comment by 7vaibhav7** [comment]
Any view on the same, @mobuchowski ? And can someone please look into incorporating this?

**Comment by mobuchowski** [comment]
Hey @7vaibhav7 I hope you're doing well! I wanted to gently remind you that our OpenSource library is maintained by contributors who respond to general requests on a voluntary manner without any promises or guarantees. While it's perfectly okay to request features or support, regular bumping of requests is not encouraged as it can be overwhelming for the maintainers, that have other responsibilities than always looking on Slack.
The best way to see something implemented is to contribute directly to the project. Your help is always appreciated, and even small contributions can make a big difference.

**Comment by mobuchowski** [comment]
@pawel-big-lebowski @JDarDagran @julienledem any thoughts on adding optional tags to those facets?

**Comment by simran-suri** [comment]
Hi, any thoughts on the same? I tried doing it but no luck, can someone please guide and help pitching in?

**Comment by simran-suri** [comment]
If not a new tag field, then is it possible to pass metadata in additionProperties field and how to get them in events?
